les/distributor.go

```diff
@@ -179,18 +184,25 @@ func (d *requestDistributor) nextRequest() (distPeer, *distReq, time.Duration) {
 	checkedPeers := make(map[distPeer]struct{})
 	elem := d.reqQueue.Front()
 	var (
-		bestPeer distPeer
-		bestReq  *distReq
 		bestWait time.Duration
 		sel      *weightedRandomSelect
 	)
 
 	d.peerLock.RLock()
 	defer d.peerLock.RUnlock()
 
-	for (len(d.peers) > 0 || elem == d.reqQueue.Front()) && elem != nil {
+	peerCount := len(d.peers)
+	for (len(checkedPeers) < peerCount || elem == d.reqQueue.Front()) && elem != nil {
```

Comment:
We should reset the `checkedPeers` if the head of request queue is removed.

Reply:
We do reset checkedPeers before the loop and we do not remove anything from the queue inside the loop, just iterate from the front to the back until we find a potential candidate for each peer. We only care about each peer until we find the first request in the queue that can be sent there and then select the global best request. After selecting and sending that request, we remove it from the queue. Next time we call nextRequest() checkedPeers will of course be reset.

Reply:
I mean in the Line 225, if the head request can't be sent, then the request will not be dropped and we start to iterate next request. In this case we should reset `checkedPeers ` but currently we don't :).

Reply:
I have the wrong understanding before, the logic here is correct.